### ISSUE TITLE
Injection?

### ISSUE THREAD

**Issue by jontonsoup** [florinpatrascu/neo4j_sips#10]
I'm wondering if this library is safe from user injection attacks?

Thanks!

**Comment by florinpatrascu** [comment]
Hi Jonathan - it depends. The driver is sending a user-given set of Neo4j (Cypher) commands to the server, being concerned about connection, transactions, pooling, errors, various transformations, etc. Therefore, if you decide to delete your nodes, let's say, the driver won't stop you. Why would it?! But why do you have this concern?! I'd imagine you'll not simply just accept external commands (free text) from anybody crafting cypher commands and just run them indiscriminately?! You'll most probably write functions (business logic) that will convert higher level business concerns i.e. Delete user, to an equivalent set of Cypher commands that would be defined into Module/named function/etc. where the user identification was already considered and so on, making sure only that user will be deleted and not other users too :) I'm over simplifying, but I believe you got the idea?! The short answer is: no, the driver is not validating the query contents, but I welcome any pull requests :) HTH - Florin

**Comment by jontonsoup** [comment]
@florinpatrascu thanks!

**Comment by sebsel** [comment]
Although it is indeed the task of the developer of the app to make sure no weird queries run... just like in SQL it's easy to make the following mistake:
```elixir
def index(conn, %{"prop" => prop}) do
  cypher =
    ~s"""
    MATCH (n)
    WHERE n.prop = "#{prop}"
    RETURN n
    """
  results = Neo4j.query!(Neo4j.conn, cypher)
  render("view.json", results)
end
```

... where `prop` is now user content, which means it can also be `x" DELETE n WHERE n.prop = "x`, for example.

The following is not currently documented (I believe?), but through a bit of trial and error I found out that it works like this:

```elixir
def index(conn, %{"prop" => prop}) do
  cypher =
    ~s"""
    MATCH (n)
    WHERE n.prop = {prop}
    RETURN n
    """
  results = Neo4j.query!(Neo4j.conn, cypher, %{"prop" => prop})
  render("view.json", results)
end
```

**Comment by florinpatrascu** [comment]
correct. If you simply interpolate any incoming params (from untrusted sources) yes, this can happen. Thank you for your comment and examples! What would you recommend, in the scenario above?

**Comment by sebsel** [comment]
I just learned more about this at a Neo4j Workshop. Turns out the interpolation is actually slower, because when using params, Neo4j can cache the query plan.

Also, my example is deprecated. The `{prop}` syntax does still work, but the new way of doing it is `$prop` in your query.

So I am recommending to always use props over interpolation. I would also recommend to update the docs to explain this feature of passing the props into the query :)
